Multihead refactor

ManMower opened the pull request:
Splitting multihead support into front and back end components

ManMower committed:
rdp: Stop using deprecated functions

The get file descriptor functions are being deprecated and a two step
process of getting handles and then getting the descriptors for the
handles is being used instead.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

michaelolbrich committed:
compositor: only reflow the outputs if the shell did not move them

weston_compositor_reflow_outputs() assumes that all output are positioned from
left to right with no gaps in the same order in which they where created.

If the shell moves an output with weston_output_move() then this assumption is
no longer true. So stop reflowing the outputs in the case. The shell is now
responsible for positioning all outputs as needed.

Signed-off-by: Michael Olbrich <m.olbrich@pengutronix.de>

ManMower committed:
rdp: Update to new FreeRDP structure layout

In an upcoming release the old style will be deprecated, so let's update
now.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
shared: Make xalloc.h stand alone

Make fail_on_null static inline and put it in xalloc.h so we can use the
header exclusively instead of having to link with the library for it.

This is so we can use xalloc in places (like the RDP backend) without
having to bring in libshared.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Refactor display management code

Split this into front/back end bits, with the front end parts putting
together a monitor array in a common form, and the back end bits in
rdpdisp.c doing the final processing.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Move RAIL ResetGraphics handling

Pulling backend handling code out to aid further refactoring.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Consolidate monitor layout change functions

Now that the ResetGraphics handling is pulled out, these are essentially
the same function. Mash them together into one. This will help later
when we push monitor layouting to the front end, as only one callback
will be required.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Move input translation code to rdpdisp.c

This will end up in the front end soon, so pull it out of the backend
include files.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Move peerlist from output to backend

Having the peer list attached to a monitor structure makes less sense now
that we have multi-head support.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Move heads to pending list properly

wl_list_insert_list() should be used for this, otherwise an empty list
becomes a loop. Currently we never see an empty list here, so it won't
break yet, but some upcoming changes will hit this.

Also gratuitously move the head_list list_init because it's conceptually
part of "moving" all the elements out of it into the pending list.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Use wl_list_for_each_safe for monitor layout change code

We can't safely remove items from a list when using wl_list_for_each,
instead we should use wl_list_for_each_safe.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Always set the preferred mode flag

This missing from non-RAIL cases is really a bug, so we can just always
set it here.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove regionWestonHeads

This is only used for printing some debug text, so let's just not keep it.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: remove regionClientHeads

We only need the desktop's bounds, we don't need to store the regions.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove regionWeston from rdp_head

This is only used to print some debug text.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove head_default from struct rdp_backend

We still need to set up regions for our single head, so push it all through
xf_peer_adjust_monitor_layout.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Don't depend on default_output in xf_input_synchronize_event

Use the first monitor in the list instead of default_monitor. In this
particular case, there will only be a single monitor, so it would be
the same.


Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Use the first monitor in rdp_peer_refresh_region

output_default would have been this anyway.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: always walk the peer list in rdp_destroy

Since the peer list has been moved to the backend, we don't have to
check b->output_default to find it. The singleton rdp_peer for RAIL is
also always in this list, so we don't need to special case it.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove dependency on output_default

We can find the primary head and its output in the main loop instead of
special casing output_default.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove output_default

We no longer need this

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Don't use rectWeston in input translation

We've set up the outputs to use these same co-ordinates, so picking them
from there puts us one step closer to eliminating head walks from
translations.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Break monitor checks into two steps

Put the basic sanity checks into one function, as these will remain in the
backend, and put the layout checks in another which will migrate to the
frontend.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove head_list from backend

The compositor already has a head list, so let's just use that instead.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove the move pending list

weston_output_move is idempotent, so it doesn't matter if we move a head
that's already in the right place. So let's just move all the heads
instead of having a special move pending list at all.

This will make things easier later when the frontend and backend code do
different parts of monitor setup later.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Rework head layout changes

Primary goal here was to have a clean split between the code that creates
and deletes heads, and the code the validates and sets modes. These
correspond to the front and and back end responsibilities respectively.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Only pass compositor to end_monitor_layout_change

When this is part of the front end, it will only have access to the
compositor structure.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: rework head layout validation

Perform this on the compositor head_list instead of on the monitor array,
which is how it needs to work to be in the front end code, where it will
only have access to the compositor list, and only after the heads have
been created.

This also moves the validation code to after the heads have been created
and removed - it validates output configuration and scale parameters, but
wasn't really changing head allocation/deallocation anyway.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Combine the two front-end output setup functions

Since these take the same parameters and always run back to back, combine
them.

At the same time convert a couple of wl_list_for_each_safe call sites that
don't remove list elements to wl_list_for_each.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove westonRect from backend structures

Change this to a local variable now that the two functions that should be
using it are combined.

This commit intentionally introduces a bug that will be fixed in a follow
up commit. Outputs may not be properly moved into place immediately
after creation.

In the future output creation and positioning will happen in the front end,
and this will no longer be a problem.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
libweston: Make weston_head_set_device_changed public

Some backends have special head specific state that doesn't fit into the
existing generic head setter functions, and is too specific to make more
functions for.

RDP's primary output flag is an example.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Pass rdp_backend to client_scale getters

We only need the backend, and we'll be calling this in places where
we don't have the peer context soon. While we're at it, make it
take an rdpMonitor as its second parameter, as this will also
help in the future.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove ClientScale from monitor struct

We can create this any time we need it, provided we appropriately change
the desktopScale factor in validation.

This lets us minimize the amount of state that needs to be passed between
front and back ends.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Stop storing scale in monitor struct

We can create this any time we need it.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove struct rdp_monitor_mode

We no longer need anything but the rdpMonitor, so let's reduce.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Store dpi settings in the front end

We'll need these for output setup

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Make backend-rdp require freerdp

We'll need some freerdp headers for monitor configuration, so include
freerdp headers from backend-rdp.h and fix up the build.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Make to_rdp_head's parameter const

It'll never change it, and some future callers care about that.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Divide head and output creation between back and front ends

Makes all the validation code part of the front end, and leaves the basic
creation in the backend.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Move some head setup to rdp_head_create

We know these at creation time, so just set them immediately.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove regionClient from head structure

We don't need a pixman region for these simple rectangles.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Remove scale only functions

They're currently both the same and perform a simple operation. For now
just use multiplication.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Move outputs in the same loop as scale and modeset

We can do this all in a single loop.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

hideyukn88 commented on libweston/backend-rdp/rdprail.c:
```diff
@@ -4123,7 +4225,7 @@ rdp_rail_notify_app_list(void *rdp_backend, struct weston_rdprail_app_list_data
 			updateAppList.flags |= RDPAPPLIST_FIELD_ICON;
 		}
-		peerCtx->applist_server_context->UpdateApplicationList(peerCtx->applist_server_context, &updateAppList);
+		peerCtx->applist_server_context->context->updateApplicationList(peerCtx->applist_server_context, &updateAppList);
```
I'm hitting compile error on this line, so do you have update on https://github.com/microsoft/wslg/blob/main/rdpapplist/rdpapplist_server.h, too?

ManMower replied:
Ouch, that was a search and replace error in a section of code that doesn't compile for me. The change is just wrong, and I've dropped it now.

hideyukn88 commented on compositor/main.c:
```diff
@@ -2768,6 +2729,392 @@ read_rdp_config_int(char *config_name, int default_value)
 	return default_value;
 }
 
+static bool
```
I think these new additions to be placed at separate file outside of main.c, like rdpdisp.c (similar idea as rdpaudio.c).

ManMower replied:
Good idea, I've moved these now.

ManMower committed:
libweston: Don't move outputs during enable

This is pretty counter-intuitive, and should probably happen outside of
the core in the front end while configuring the outputs.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
libweston: Check output placement

Make sure we don't enable an output that overlaps with other enabled
outputs.

We should probably do something similar when moving outputs, but we can't
realistically do that right now, so at least leave a comment explaining
why we're ignoring that case.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
libweston: Don't send output_changed signal when moving disabled outputs

weston_output_set_position() currently assumes the output is enabled, but
we could be using weston_output_move() to configure an output that hasn't
yet been enabled.

If that's the case, we don't want to send signals or perform setup that
will eventually happen when the output is enabled anyway.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

ManMower committed:
rdp: Add a perfect match for the head re-use test

It may be better in some cases to look for a perfect match, and the
search is fairly cheap, so let's throw it in the mix.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>

hideyukn88 commented on include/libweston/backend-rdp.h:
```diff
@@ -33,21 +33,19 @@ extern "C" {
 #include <libweston/libweston.h>
 #include <libweston/plugin-registry.h>
 
+#include <freerdp/freerdp.h>
```
This is a bit concern as anyone includes this header needs to have dependencies to FreeRDP, like rdprail-shell in this PR. Should we have our own rdp monitor struct rather than using same from FreeRDP?

ManMower replied:
I think it's OK to require this dependency for anyone that needs to include the backend-rdp header. If they're including our backend-rdp.h header, there's some expectation that they may have to deal with FreeRDP specifics.

fooishbar replied:
Yeah. It would be great if FreeRDP had a more stable API, but given that frontends are going to have to deal with it anyway, I don't think there's any value in defining our own wrapper struct. After all, if libweston can create an rdpMonitor equivalent which expresses everything required & is ABI-stable - why couldn't that struct just be provided by FreeRDP itself?

hideyukn88 replied:
> I think it's OK to require this dependency for anyone that needs to include the backend-rdp header. If they're including our backend-rdp.h header, there's some expectation that they may have to deal with FreeRDP specifics.

But that's not true for rdprail-shell, they don't access any FreeRDP API or struct, but now because it's including backend-rdp header, it has to take dependency to FreeRDP.

fooishbar replied:
There are definitely some very simple shells who wouldn't require it for sure. But when you get into providing more involved functionality like audio, clipboard, app IDs, etc, then the frontend is required (after some of these PRs) to deal with RDP directly.

hideyukn88 replied:
OK, then keep as-is, thanks for the response.

hideyukn88 commented on rdprail-shell/meson.build:
```diff
@@ -35,6 +35,7 @@ if get_option('shell-rdprail')
 		dep_librsvg,
 		dep_winpr,
 		dep_glib,
+		dep_frdp_server,
```
Ideally we don't need this if backend-rdp header doesn't include FreeRDP header.

hideyukn88 commented on libweston/backend-rdp/rdp.c:
```diff
@@ -370,182 +382,81 @@ ensure_matching_mode(struct weston_output *output, struct weston_mode *target)
 
-		/* In HiDefRemoteApp mode, free previous current_mode,
-		   since it only want to expose current mode to app */
-		wl_list_remove(&previous_mode->link);
```
As I replied in email, we still need this, so "available mode" will be only current mode.

ManMower replied:
Yes, your'e right - I've added a new commit to address this in a slightly different way.

ManMower committed:
rdp: Only keep one mode on the mode list

RDP outputs only really have a single mode - either the mode set in
configuration, or the mode set at connection time. Make sure we only have
the one mode in the mode list, and update it when appropriate.

Signed-off-by: Derek Foreman <derek.foreman@collabora.com>
